Adapt downsteam Ghost Squid support to latest QMK

fenuks opened the pull request:
## Description
This changeset adapts code from https://github.com/BathroomEpiphanies/epiphanies_qmk_keyboard/tree/master/keyboards/ghost_squid_20140518 to add upstream support for Ghost Squid controller for CM Storm QuickFire XT. I didn't write that code, I just adapted it to compiler under the latest QMK as mentioned in #3952.

This is still a draft, as documentation is lacking. I will add it, but I thought it would be useful to get prior feedback before I create info.json and KLE definition. Note that I'm new to QMK, so probably there is much room for improvement. I mostly changed keymap code, as FN keys didn't work at all. I removed completely blowrak keymap found in original repo to make things simpler for me (it's, citing creator, `Swedish take on Dvorak`). I expect that more code can be deleted/cleaned.

## Types of Changes
- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [X] Keyboard (addition or update)
- [X] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* #3952

## Checklist

- [X] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [X] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [X] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [X] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [X] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

drashna commented on keyboards/bpiphany/ghost_squid/config.h:
```diff
@@ -0,0 +1,42 @@
+/* Debounce reduces chatter (unintended double-presses) - set 0 if debouncing is not needed */
+#define DEBOUNCE 5
+
+/* key combination for magic key command */
+#define IS_COMMAND() ( \
+	keyboard_report->mods == (MOD_BIT(KC_LSHIFT) | MOD_BIT(KC_RSHIFT)) \
+)
```
```suggestion
```

drashna commented on keyboards/bpiphany/ghost_squid/ghost_squid.h:
```diff
@@ -0,0 +1,54 @@
+
+#pragma once
+
+#include "matrix.h"
+#include "quantum.h"
+
+inline void num_led_on(void)  { DDRC |=  (1<<5); PORTC &= ~(1<<5); }
+inline void num_led_off(void) { DDRC &= ~(1<<5); PORTC &= ~(1<<5); }
+
+inline void caps_led_on(void)  { DDRC |=  (1<<6); PORTC &= ~(1<<6); }
+inline void caps_led_off(void) { DDRC &= ~(1<<6); PORTC &= ~(1<<6); }
+
+inline void scroll_led_on(void)  { DDRB |=  (1<<7); PORTB &= ~(1<<7); }
+inline void scroll_led_off(void) { DDRB &= ~(1<<7); PORTB &= ~(1<<7); }
+
+inline void fn_led_on(void)  { DDRD |=  (1<<0); PORTD &= ~(1<<0); }
+inline void fn_led_off(void) { DDRD &= ~(1<<0); PORTD &= ~(1<<0); }
```
```suggestion
#include "quantum.h"

#define fn_led_on() writePinHigh(D0)
#define fn_led_off() writePinLow(D0)
```

drashna commented on keyboards/bpiphany/ghost_squid/ghost_squid.c:
```diff
@@ -0,0 +1,80 @@
+void matrix_scan_kb(void) {
+	matrix_scan_user();
+}
+
+bool process_action_kb(keyrecord_t *record) {
+	return process_action_user(record);
+}
+
+void led_set_kb(uint8_t usb_led) {
+	led_set_user(usb_led);
+}
+
+void led_set_user(uint8_t usb_led) {
+	if (usb_led & (1 << USB_LED_NUM_LOCK)) {
+		num_led_on();
+	} else {
+		num_led_off();
+	}
+
+	if (usb_led & (1 << USB_LED_CAPS_LOCK)) {
+		caps_led_on();
+	} else {
+		caps_led_off();
+	}
+
+	if (usb_led & (1 << USB_LED_SCROLL_LOCK)) {
+		scroll_led_on();
+	} else {
+		scroll_led_off();
+	}
+}
+
+extern inline void num_led_on(void);
+extern inline void num_led_off(void);
+extern inline void caps_led_on(void);
+extern inline void caps_led_off(void);
+extern inline void scroll_led_on(void);
+extern inline void scroll_led_off(void);
+extern inline void fn_led_on(void);
+extern inline void fn_led_off(void);
```
```suggestion
void keyboard_pre_init_kb(void) {
    setPinOutput(D0);
    writePinLow(D0);
    
    keyboard_pre_init_user();
}
```

drashna commented on keyboards/bpiphany/ghost_squid/keymaps/default/rules.mk:
```diff
@@ -0,0 +1,17 @@
+# Build Options
+#   change to "no" to disable the options, or define them in the Makefile in 
+#   the appropriate keymap folder that will get included automatically
+#
+BOOTMAGIC_ENABLE ?= yes      # Virtual DIP switch configuration(+1000)
+MOUSEKEY_ENABLE ?= yes       # Mouse keys(+4700)
+EXTRAKEY_ENABLE ?= yes       # Audio control and System control(+450)
+CONSOLE_ENABLE ?= yes        # Console for debug(+400)
+COMMAND_ENABLE ?= yes        # Commands for debug and configuration
+CUSTOM_MATRIX ?= yes         # Custom matrix file for the Pegasus Hoof due to the 2x74HC42
+NKRO_ENABLE ?= no            # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE ?= no       # Enable keyboard backlight functionality
+MIDI_ENABLE ?= no            # MIDI controls
+AUDIO_ENABLE ?= no           # Audio output on port C6
+UNICODE_ENABLE ?= no         # Unicode
+BLUETOOTH_ENABLE ?= no       # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE ?= no        # Enable WS2812 RGB underlight.  Do not enable this with audio at the same time.
```
This file should be removed, or only the features that are not turned on at the kb level left.

```suggestion
```

drashna commented on keyboards/bpiphany/ghost_squid/matrix.c:
```diff
@@ -0,0 +1,167 @@
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <stdint.h>
```
This should be converted to matrix lite:

https://docs.qmk.fm/#/custom_matrix?id=lite

drashna commented on keyboards/bpiphany/ghost_squid/rules.mk:
```diff
@@ -0,0 +1,66 @@
+#     calculate timings. Do NOT tack on a 'UL' at the end, this will be done
+#     automatically to create a 32-bit value in your source code.
+#
+#     This will be an integer division of F_USB below, as it is sourced by
+#     F_USB after it has run through any CPU prescalers. Note that this value
+#     does not *change* the processor frequency - it should merely be updated to
+#     reflect the processor speed set externally so that the code can use accurate
+#     software delays.
+F_CPU = 16000000
+
+#
+# LUFA specific
+#
+# Target architecture (see library "Board Types" documentation).
+ARCH = AVR8
+
+# Input clock frequency.
+#     This will define a symbol, F_USB, in all source code files equal to the
+#     input clock frequency (before any prescaling is performed) in Hz. This value may
+#     differ from F_CPU if prescaling is used on the latter, and is required as the
+#     raw input clock is fed directly to the PLL sections of the AVR for high speed
+#     clock generation for the USB and other AVR subsections. Do NOT tack on a 'UL'
+#     at the end, this will be done automatically to create a 32-bit value in your
+#     source code.
+#
+#     If no clock division is performed on the input clock inside the AVR (via the
+#     CPU clock adjust registers or the clock division fuses), this will be equal to F_CPU.
+F_USB = $(F_CPU)
+
+# Interrupt driven control endpoint task(+60)
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
+
+# Boot Section Size in *bytes*
+#   Teensy halfKay   512
+#   Teensy++ halfKay 1024
+#   Atmel DFU loader 4096
+#   LUFA bootloader  4096
+#   USBaspLoader     2048
+OPT_DEFS += -DBOOTLOADER_SIZE=4096
```
```suggestion
# Bootloader selection
BOOTLOADER = atmel-dfu
```

drashna commented on keyboards/bpiphany/ghost_squid/rules.mk:
```diff
@@ -0,0 +1,66 @@
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE ?= yes      # Virtual DIP switch configuration(+1000)
+MOUSEKEY_ENABLE ?= yes       # Mouse keys(+4700)
+EXTRAKEY_ENABLE ?= yes       # Audio control and System control(+450)
+CONSOLE_ENABLE ?= yes        # Console for debug(+400)
+COMMAND_ENABLE ?= yes        # Commands for debug and configuration
+NKRO_ENABLE ?= no            # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE ?= no       # Enable keyboard backlight functionality
+AUDIO_ENABLE ?= no           # Audio output on port C6
+UNICODE_ENABLE ?= no         # Unicode
+BLUETOOTH_ENABLE ?= no       # Enable Bluetooth with the Adafruit EZ-Key HID
+RGBLIGHT_ENABLE ?= no        # Enable WS2812 RGB underlight.  Do not enable this with audio at the same time.
+
+
+CUSTOM_MATRIX ?= yes
+SRC += matrix.c
```
```suggestion
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
NKRO_ENABLE = no            # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
AUDIO_ENABLE = no           # Audio output
RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight.  Do not enable this with audio at the same time.

CUSTOM_MATRIX = yes
SRC += matrix.c
```

drashna commented on keyboards/bpiphany/ghost_squid/config.h:
```diff
@@ -0,0 +1,42 @@
+/* COL2ROW or ROW2COL */
+#define DIODE_DIRECTION COL2ROW
+
+/* Debounce reduces chatter (unintended double-presses) - set 0 if debouncing is not needed */
```
```suggestion
#define LED_NUM_LOCK_PIN C5
#define LED_CAPS_LOCK_PIN C6
#define LED_SCROLL_LOCK_PIN B7

/* Debounce reduces chatter (unintended double-presses) - set 0 if debouncing is not needed */
```

fenuks commented:
Thank you for your suggestions. I applied almost all, I still read about matrix lite. I had to leave `matrix_init_kb` and `matrix_scan_kb`, otherwise code wouldn't link. Besides that, I added another layer to disable super/gui keys (keyboard allows to disable those when Fn + F9 is pressed).

fenuks commented:
I've been trying to convert code to matrix lite, but it doesn't work, i.e. I cannot type after I flash new firmware. My apologies if I'm asking about something obvious, but after browsing over many other matrix.c files I didn't find hints how to solve it.

I think that problem might lie in fact currently custom `matrix_print` and `matrix_key_count` act as if MATRIX_COLS <= 16, but in config.h it is defines as `#define MATRIX_COLS 18`, thus default implementation in [matrix_common.c](https://github.com/qmk/qmk_firmware/blob/master/quantum/matrix_common.c#L63) will print different header. Same for key count, bitpop32 will be used instead of bitpop16. I don't understand codebase well yet, so I'm not certain if that's the root of problem. Other than that, it seems `matrix_is_modified` also needs to be rewritten, because it works differently in matrix_common.c implementation.

fauxpark commented:
For Custom Matrix Lite, a lot of that stuff is not necessary as it's already implemented in core. At minimum you only need to implement the init and scan functions.

fenuks commented:
I tried to do that, I removed all the functions that have provided default implementation, and renamed appropriately init and scan functions, but it doesn't work. Key presses are not detected at all. I don't know how important are `matrix_print` and `matrix_key_count`, but it seems to me that in default definition they work differently because custom definitions act as if MATRIX_COLUNS was <= 16.

fauxpark commented on keyboards/bpiphany/ghost_squid/ghost_squid.c:
```diff
@@ -0,0 +1,30 @@
+
+#include "ghost_squid.h"
+
+void matrix_init_kb(void) {}
+
+void matrix_scan_kb(void) {}
+
```
```suggestion
```

fauxpark commented on keyboards/bpiphany/ghost_squid/keymaps/default/config.h:
```diff
@@ -0,0 +1,3 @@
+#pragma once
```
This file can be deleted.

fenuks replied:
I added it, because I noticed that `frosty_flake` includes such file; removed.

fauxpark commented on keyboards/bpiphany/ghost_squid/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,75 @@
+* └────┴────┴────┴────────────────────────┴────┴────┴────┴────┘ └───┴───┴───┘ └───────┴───┴───┘
+*/
+
+#define KM_QWERTY     0
+#define KM_MEDIA      1
+#define KM_GUI_LOCK   2
```
```suggestion
enum layer_names {
    KM_QWERTY,
    KM_MEDIA,
    KM_GUI_LOCK
};
```

fauxpark commented on keyboards/bpiphany/ghost_squid/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,75 @@
+KC_LCTL, KC_LGUI,KC_LALT,                       KC_SPC,                          KC_RALT,KC_RGUI,MO(KM_MEDIA),KC_RCTL,   KC_LEFT,KC_DOWN,KC_RGHT,    KC_P0,          KC_PDOT),
+	/* Layer 1: Function layer */
+	[KM_MEDIA] = LAYOUT(
+_______, _______,  _______,  _______,  _______,  KC_MEDIA_PLAY_PAUSE, KC_MEDIA_STOP, KC_MEDIA_PREV_TRACK, KC_MEDIA_NEXT_TRACK,  TG(KM_GUI_LOCK),KC_MUTE, KC_VOLD, KC_VOLU,   _______,_______, RESET, \
+_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,    _______,_______,_______,_______, \
+		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,   _______,_______,_______,    _______,  _______,  _______,  _______, \
+		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,        _______,                               _______,  _______,  _______,           \
+		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,        _______,           _______,            _______,  _______,  _______,  _______, \
```
```suggestion
_______, _______,  _______,  _______,  _______,  KC_MEDIA_PLAY_PAUSE, KC_MEDIA_STOP, KC_MEDIA_PREV_TRACK, KC_MEDIA_NEXT_TRACK,  TG(KM_GUI_LOCK),KC_MUTE, KC_VOLD, KC_VOLU,   _______,_______, RESET,
_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,    _______,_______,_______,_______,
		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,   _______,_______,_______,    _______,  _______,  _______,  _______,
		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,        _______,                               _______,  _______,  _______,
		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,        _______,           _______,            _______,  _______,  _______,  _______,
```

fauxpark commented on keyboards/bpiphany/ghost_squid/keymaps/default/keymap.c:
```diff
@@ -0,0 +1,75 @@
+		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,        _______,           _______,            _______,  _______,  _______,  _______, \
+		_______,_______,_______,                        _______,                        _______,_______,_______  ,_______,   _______,_______,_______,    _______,          _______),
+	[KM_GUI_LOCK] = LAYOUT(
+_______, _______,  _______,  _______,  _______,  _______, _______, _______, _______,  _______,_______, _______, _______,   _______,_______, _______, \
+_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,    _______,_______,_______,_______, \
+		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,   _______,_______,_______,    _______,  _______,  _______,  _______, \
+		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,        _______,                               _______,  _______,  _______,           \
+		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,        _______,           _______,            _______,  _______,  _______,  _______, \
```
```suggestion
_______, _______,  _______,  _______,  _______,  _______, _______, _______, _______,  _______,_______, _______, _______,   _______,_______, _______,
_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,    _______,_______,_______,_______,
		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,   _______,_______,_______,    _______,  _______,  _______,  _______,
		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,        _______,                               _______,  _______,  _______,
		_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,_______,        _______,           _______,            _______,  _______,  _______,  _______,
```

fauxpark commented on keyboards/bpiphany/ghost_squid/readme.md:
````diff
@@ -0,0 +1,34 @@
+
+Make example for this keyboard (after setting up your build environment):
+
+104 key default layout:
+
+```
+make bpiphany/ghost_squid:default
+```
````
```suggestion
    make bpiphany/ghost_squid:default
```

fauxpark commented on keyboards/bpiphany/ghost_squid/readme.md:
```diff
@@ -0,0 +1,34 @@
+The code was adapted from the [BathroomEpiphanies QMK
+Firmware](https://github.com/BathroomEpiphanies/epiphanies_qmk_keyboard/tree/master/keyboards/ghost_squid_20140518).
+
+## Layout support
+
+Support for both 104 key layout is provided. See the
+keymaps `default` for example layout.
+
+Keyboard Maintainer: QMK Community
+Hardware Supported: Ghost Squid
+Hardware Availability: [1upkeyboards](https://1upkeyboards.com/shop/controllers/qf-xt-ghost-squid-controller-2/)
```
```suggestion
* Keyboard Maintainer: QMK Community
* Hardware Supported: Ghost Squid
* Hardware Availability: [1upkeyboards](https://1upkeyboards.com/shop/controllers/qf-xt-ghost-squid-controller-2/)
```

fauxpark commented on keyboards/bpiphany/ghost_squid/readme.md:
````diff
@@ -0,0 +1,34 @@
+make bpiphany/ghost_squid:default
+```
+
+To directly flash the ghost_squid after compiling use:
+
+```
+make bpiphany/ghost_squid:default:dfu
+```
````
```suggestion
Flashing example for this keyboard:

    make bpiphany/ghost_squid:default:flash
```

fauxpark commented on keyboards/bpiphany/ghost_squid/readme.md:
```diff
@@ -0,0 +1,34 @@
+# Ghost Squid Controller
+
+This is the firmware for Rev. 20140518 of the Ghost Squid controller
+by [Bathroom Epiphanies](http://bathroomepiphanies.com/controllers/),
+a replacement controller for the [Cooler Master Quick Fire
+XT](https://www.coolermaster.com/catalog/peripheral/keyboards/quick-fire-xt/).
+
+The code was adapted from the [BathroomEpiphanies QMK
+Firmware](https://github.com/BathroomEpiphanies/epiphanies_qmk_keyboard/tree/master/keyboards/ghost_squid_20140518).
```
```suggestion
This is the firmware for Rev. 20140518 of the Ghost Squid controller by [Bathroom Epiphanies](http://bathroomepiphanies.com/controllers/), a replacement controller for the [Cooler Master Quick Fire XT](https://www.coolermaster.com/catalog/peripheral/keyboards/quick-fire-xt/).

The code was adapted from the [BathroomEpiphanies QMK Firmware](https://github.com/BathroomEpiphanies/epiphanies_qmk_keyboard/tree/master/keyboards/ghost_squid_20140518).
```

fauxpark commented on keyboards/bpiphany/ghost_squid/rules.mk:
```diff
@@ -0,0 +1,19 @@
+# MCU name
+MCU = atmega32u2
+
+BOOTLOADER = atmel-dfu
```
```suggestion
# Bootloader selection
BOOTLOADER = atmel-dfu
```

fauxpark commented on keyboards/bpiphany/ghost_squid/rules.mk:
```diff
@@ -0,0 +1,19 @@
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
+MOUSEKEY_ENABLE = yes       # Mouse keys
+EXTRAKEY_ENABLE = yes       # Audio control and System control
+CONSOLE_ENABLE = no         # Console for debug
+COMMAND_ENABLE = no         # Commands for debug and configuration
+NKRO_ENABLE = no            # Nkey Rollover - if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
+AUDIO_ENABLE = no           # Audio output
+RGBLIGHT_ENABLE = no        # Enable WS2812 RGB underlight.  Do not enable this with audio at the same time.
+CUSTOM_MATRIX = yes
+SRC += matrix.c
```
```suggestion
BOOTMAGIC_ENABLE = yes      # Enable Bootmagic Lite
MOUSEKEY_ENABLE = yes       # Mouse keys
EXTRAKEY_ENABLE = yes       # Audio control and System control
CONSOLE_ENABLE = no         # Console for debug
COMMAND_ENABLE = no         # Commands for debug and configuration
# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
NKRO_ENABLE = no            # USB Nkey Rollover
BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality
RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
AUDIO_ENABLE = no           # Audio output
CUSTOM_MATRIX = lite

SRC += matrix.c
```

fauxpark commented on keyboards/bpiphany/ghost_squid/matrix.c:
```diff
@@ -0,0 +1,167 @@
+  }
+  return count;
+}
+
+static matrix_row_t read_cols(void)
+{
+  return
+    (PINB&(1<<1) ? 0 : ((matrix_row_t)1<<0)) |
+    (PINC&(1<<2) ? 0 : ((matrix_row_t)1<<1)) |
+    (PINB&(1<<6) ? 0 : ((matrix_row_t)1<<2)) |
+    (PINB&(1<<4) ? 0 : ((matrix_row_t)1<<3)) |
+    (PINB&(1<<3) ? 0 : ((matrix_row_t)1<<4)) |
+    (PINB&(1<<2) ? 0 : ((matrix_row_t)1<<5)) |
+    (PINB&(1<<0) ? 0 : ((matrix_row_t)1<<6)) |
+    (PINB&(1<<5) ? 0 : ((matrix_row_t)1<<7)); 
+}
+
+static void select_row(uint8_t col)
+{
+  switch (col) {
+  case  0: PORTD = (PORTD & ~0b01111110) | 0b01100010; break;
+  case  1: PORTD = (PORTD & ~0b01111110) | 0b01101000; break;
+  case  2: PORTD = (PORTD & ~0b01111110) | 0b01101100; break;
+  case  3: PORTD = (PORTD & ~0b01111110) | 0b01110000; break;
+  case  4: PORTD = (PORTD & ~0b01111110) | 0b01111000; break;
+  case  5: PORTD = (PORTD & ~0b01111110) | 0b01100000; break;
+  case  6: PORTD = (PORTD & ~0b01111110) | 0b01110100; break;
+  case  7: PORTD = (PORTD & ~0b01111110) | 0b01100100; break;
+  case  8: PORTD = (PORTD & ~0b01111110) | 0b01111100; break;
+  case  9: PORTD = (PORTD & ~0b01111110) | 0b01101010; break;
+  case 10: PORTD = (PORTD & ~0b01111110) | 0b00110110; break;
+  case 11: PORTD = (PORTD & ~0b01111110) | 0b00010110; break;
+  case 12: PORTD = (PORTD & ~0b01111110) | 0b01001110; break;
+  case 13: PORTD = (PORTD & ~0b01111110) | 0b00111110; break;
+  case 14: PORTD = (PORTD & ~0b01111110) | 0b00011110; break;
+  case 15: PORTD = (PORTD & ~0b01111110) | 0b01000110; break;
+  case 16: PORTD = (PORTD & ~0b01111110) | 0b00100110; break;
+  case 17: PORTD = (PORTD & ~0b01111110) | 0b00101110; break;
+  }
+}
```
```suggestion
matrix_row_t read_rows(void) {
    return
        (PINB & (1 << 1) ? 0 : ((matrix_row_t)1 << 0)) |
        (PINC & (1 << 2) ? 0 : ((matrix_row_t)1 << 1)) |
        (PINB & (1 << 6) ? 0 : ((matrix_row_t)1 << 2)) |
        (PINB & (1 << 4) ? 0 : ((matrix_row_t)1 << 3)) |
        (PINB & (1 << 3) ? 0 : ((matrix_row_t)1 << 4)) |
        (PINB & (1 << 2) ? 0 : ((matrix_row_t)1 << 5)) |
        (PINB & (1 << 0) ? 0 : ((matrix_row_t)1 << 6)) |
        (PINB & (1 << 5) ? 0 : ((matrix_row_t)1 << 7));
}

void select_col(uint8_t col) {
    switch (col) {
        case  0: PORTD = (PORTD & ~0b01111110) | 0b01100010; break;
        case  1: PORTD = (PORTD & ~0b01111110) | 0b01101000; break;
        case  2: PORTD = (PORTD & ~0b01111110) | 0b01101100; break;
        case  3: PORTD = (PORTD & ~0b01111110) | 0b01110000; break;
        case  4: PORTD = (PORTD & ~0b01111110) | 0b01111000; break;
        case  5: PORTD = (PORTD & ~0b01111110) | 0b01100000; break;
        case  6: PORTD = (PORTD & ~0b01111110) | 0b01110100; break;
        case  7: PORTD = (PORTD & ~0b01111110) | 0b01100100; break;
        case  8: PORTD = (PORTD & ~0b01111110) | 0b01111100; break;
        case  9: PORTD = (PORTD & ~0b01111110) | 0b01101010; break;
        case 10: PORTD = (PORTD & ~0b01111110) | 0b00110110; break;
        case 11: PORTD = (PORTD & ~0b01111110) | 0b00010110; break;
        case 12: PORTD = (PORTD & ~0b01111110) | 0b01001110; break;
        case 13: PORTD = (PORTD & ~0b01111110) | 0b00111110; break;
        case 14: PORTD = (PORTD & ~0b01111110) | 0b00011110; break;
        case 15: PORTD = (PORTD & ~0b01111110) | 0b01000110; break;
        case 16: PORTD = (PORTD & ~0b01111110) | 0b00100110; break;
        case 17: PORTD = (PORTD & ~0b01111110) | 0b00101110; break;
    }
}

void matrix_init_custom(void) {
    /* Column output pins */
    setPinOutput(D1);
    setPinOutput(D2);
    setPinOutput(D3);
    setPinOutput(D4);
    setPinOutput(D5);
    setPinOutput(D6);

    /* Row input pins */
    writePinHigh(B0);
    writePinHigh(B1);
    writePinHigh(B2);
    writePinHigh(B3);
    writePinHigh(B4);
    writePinHigh(B5);
    writePinHigh(B6);
    writePinHigh(C2);
}

bool matrix_scan(matrix_row_t current_matrix[]) {
    bool changed = false;

    for (uint8_t col = 0; col < MATRIX_COLS; col++) {
        select_col(col);
        matrix_io_delay();
        matrix_row_t rows = read_rows();

        for (uint8_t row = 0; row < MATRIX_ROWS; row++) {
            bool prev_bit = current_matrix[row] & ((matrix_row_t)1 << col);
            bool curr_bit = rows & (1 << row);

            if (prev_bit != curr_bit) {
                current_matrix[row] ^= (matrix_row_t)1 << col;
                changed = true;
            }
        }
    }

    return changed;
}
```
Not 100% sure about this, but might work. There's still some stuff to be cleaned up.

fenuks commented:
@fauxpark  Thank you very much for your help. Your matrix code works perfectly well. I applied other suggestions as well.

Beside more code cleaning, is there anything else you would like to be changed? I will get now to documentation, specifically diagram as requested in documentation template. Please let me know if I forgot about something.

fenuks commented:
I created image to visualize layers. https://imgur.com/a/zYIGf7L

If it's okay, I will update readme with short description.

drashna commented on keyboards/bpiphany/ghost_squid/readme.md:
```diff
@@ -0,0 +1,21 @@
+* Hardware Supported: Ghost Squid
+* Hardware Availability: [1upkeyboards](https://1upkeyboards.com/shop/controllers/qf-xt-ghost-squid-controller-2/)
+
+Make example for this keyboard (after setting up your build environment):
```
Could you add instructions on how to enter the bootloader, to flash new firmware? 

Eg: 
```markup
## Bootloader

Enter the bootloader in 3 ways:

* **Bootmagic reset**: Hold down the key at (0,0) in the matrix (usually the top left key or Escape) and plug in the keyboard
* **Physical reset button**: Briefly press the button on the back of the PCB - some may have pads you must short instead
* **Keycode in layout**: Press the key mapped to `RESET` if it is available
```

fenuks replied:
Thanks, added.
